### PR TITLE
event serializer sample

### PR DESCRIPTION
**Pull request by samarabbas**
## Checklist

1. Closes issue: 

2. How was this tested:

3. Any docs updates needed?

### CODE REVIEW

File: serializer/processor.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		ctx     workflow.Context
+		logger  log.Logger
+		eventCh workflow.ReceiveChannel
+
+		state *EventProcessorState
+	}
+
+	EventProcessorState struct {
+		PreviousEventID int
+		Items           map[int]*ResourceEvent
```
Comment (mfateev) [inline serializer/processor.go:19]: Consider not using pointers for small structures.

File: serializer/processor.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	doneCh.Send(p.ctx, p.state)
+}
+
+func (p *EventProcessor) processInOrder() {
+	s := p.state
+	nextID := s.PreviousEventID + 1
+	for current, ok := s.Items[nextID]; ok; current, ok = s.Items[nextID] {
+		p.logger.Info("Processing event",
+			"EventID", current.EventID)
+		err := workflow.ExecuteActivity(p.ctx, ProcessEvent, current).Get(p.ctx, nil)
```
Comment (mfateev) [inline serializer/processor.go:77]: Using context which was created by a different goroutine is a bug.

File: serializer/worker/main.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		log.Fatal("Unable to create client")
+	}
+	defer c.Close()
+
+	ctx := context.Background()
+	w := worker.New(c, serializer.Task_Queue_Name, worker.Options{
+		BackgroundActivityContext: ctx,
+	})
+
+	w.RegisterWorkflowWithOptions(serializer.ResourceWorkflow, workflow.RegisterOptions{Name: "ResourceWorkflow"})
```
Comment (mfateev) [inline serializer/worker/main.go:30]: I would remove the whole Name thing. Just:
```
w.RegisterWorkflow(serializer.ResourceWorkflow)
```

File: serializer/workflow.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	"time"
+
+	"go.temporal.io/sdk/activity"
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	NextRunParams struct {
+		PreviousEventID int
+		Items           map[int]*ResourceEvent
```
Comment (mfateev) [inline serializer/workflow.go:15]: Why pointer?

File: serializer/workflow.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	)
+
+	// Pump to continue processing events until the close timer fires
+	for !closeWorkflow {
+		selector.Select(ctx)
+	}
+
+	// Prepare to close this workflow execution
+
+	// First drain all unprocessed signals
```
Comment (mfateev) [inline serializer/workflow.go:94]: You can eliminate this by using `selector.HasPending()`

File: serializer/processor.go (hunk 2 above)
Comment (samarabbas) [reply serializer/processor.go:77]: This is the context created by the same goroutine

File: serializer/workflow.go (hunk 4 above)
Comment (samarabbas) [reply serializer/workflow.go]: Fixed.

File: serializer/workflow.go (hunk 5 above)
Comment (samarabbas) [reply serializer/workflow.go]: Done.

File: serializer/processor.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
+package serializer
+
+import (
+	"go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	EventProcessor struct {
+		ctx     workflow.Context
```
[the remark this answers is not in the document]
Comment (samarabbas) [reply serializer/processor.go]: Makes sense.  Removed as a field.

File: serializer/processor.go (hunk 1 above)
Comment (samarabbas) [reply serializer/processor.go]: Done.

File: serializer/worker/main.go (hunk 3 above)
Comment (samarabbas) [reply serializer/worker/main.go]: Done.

File: serializer/processor.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		}
+
+		p.state.addEvent(event, p.logger)
+		p.processInOrder(ctx)
+	}
+
+	doneCh.Send(ctx, p.state)
+}
+
+func (p *EventProcessor) processInOrder(ctx workflow.Context) {
```
Comment (mfateev) [inline serializer/processor.go:71]: This function feels like belonging to EventProcessorState.

File: serializer/workflow.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	// Prepare to close this workflow execution
+
+	// Close event processor
+	ch.Close()
+
+	// Wait for processor to completely shutdown and return state
+	var processorState *EventProcessorState
+	processorShutdownCh.Receive(ctx, &processorState)
```
Comment (mfateev) [inline serializer/workflow.go:99]: I'm not sure, but I have a hunch that adding `processorShutdownCh` to the selector above would eliminate the drain down in the code.

**Review by mfateev — APPROVED** [review]

**Pull request closed without merging**